gunrock/app/kcore/kcore_problem.cuh

```diff
@@ -0,0 +1,382 @@
+    util::Array1D<SizeT, int> degrees;
+    util::Array1D<SizeT, SizeT> k_cores;
+    util::Array1D<SizeT, VertexT> initial_frontier;
+    util::Array1D<SizeT, SizeT> empty_flag;
+    util::Array1D<unsigned int, unsigned int> delete_bitmap;
+    util::Array1D<unsigned int, unsigned int> to_be_deleted_bitmap;
+    // </TODO>
+
+    /*
+     * @brief Default constructor
+     */
+    DataSlice() : BaseDataSlice() {
+      // <TODO> name of the problem specific arrays:
+      degrees.SetName("degrees");
+      k_cores.SetName("k cores");
+      initial_frontier.SetName("array to reset frontier");
+      empty_flag.SetName("flag for whether graph is empty");
+      delete_bitmap.SetName("bitmap of vertex deleted status");
+      to_be_deleted_bitmap.SetName("bitmap of vertices marked for deletion");
+      // </TODO>
+    }
+
+    /*
+     * @brief Default destructor
+     */
+    virtual ~DataSlice() { Release(); }
+
+    /*
+     * @brief Releasing allocated memory space
+     * @param[in] target      The location to release memory from
+     * \return    cudaError_t Error message(s), if any
+     */
+    cudaError_t Release(util::Location target = util::LOCATION_ALL) {
+      cudaError_t retval = cudaSuccess;
+      if (target & util::DEVICE) GUARD_CU(util::SetDevice(this->gpu_idx));
+
+      // <TODO> Release problem specific data, e.g.:
+      GUARD_CU(degrees.Release(target));
+      GUARD_CU(k_cores.Release(target));
+      GUARD_CU(initial_frontier.Release(target));
+      GUARD_CU(empty_flag.Release(target));
+      GUARD_CU(delete_bitmap.Release(target));
+      GUARD_CU(to_be_deleted_bitmap.Release(target));
+      // </TODO>
+
+      GUARD_CU(BaseDataSlice ::Release(target));
+      return retval;
+    }
+
+    /**
+     * @brief initializing k-core-specific data on each gpu
+     * @param     sub_graph   Sub graph on the GPU.
+     * @param[in] gpu_idx     GPU device index
+     * @param[in] target      Targeting device location
+     * @param[in] flag        Problem flag containling options
+     * \return    cudaError_t Error message(s), if any
+     */
+    cudaError_t Init(GraphT &sub_graph, int num_gpus, int gpu_idx,
+                     util::Location target, ProblemFlag flag) {
+      cudaError_t retval = cudaSuccess;
+
+      GUARD_CU(BaseDataSlice::Init(sub_graph, num_gpus, gpu_idx, target, flag));
+
+      // <TODO> allocate problem specific data here, e.g.:
+      GUARD_CU(degrees.Allocate(sub_graph.nodes, target));
+      GUARD_CU(k_cores.Allocate(sub_graph.nodes, target));
+      GUARD_CU(initial_frontier.Allocate(sub_graph.nodes, target));
+      GUARD_CU(empty_flag.Allocate(1, target));
+      GUARD_CU(delete_bitmap.Allocate(((sub_graph.nodes / 32) + 1), target));
+      GUARD_CU(to_be_deleted_bitmap.Allocate(((sub_graph.nodes / 32) + 1), target));
+      // </TODO>
+
+      GUARD_CU2(cudaDeviceSynchronize(), "cudaDeviceSynchronize failed.");
+
+      if (target & util::DEVICE) {
```

Comment:
What if the graph is already allocated on the GPU when Problem::Init is called, but `target == util::DEVICE`? You can see it handled in hits_problem.cuh: https://github.com/gunrock/gunrock/blob/a7fc6948f397912ca0c8f1a8ccf27d1e9677f98f/gunrock/app/hits/hits_problem.cuh#L174

Reply:
Done, hopefully correctly? https://github.com/gunrock/gunrock/pull/842/commits/8dea4711205218851e6e9ef731c29ff6fdf74091